Move and rewrite intro pattern section

jfehrle opened the pull request:
**Kind:** documentation.

Fixes / closes https://github.com/coq/coq/issues/9221

This is a work in progress.  It needs a fair bit of help from those who know more.

This PR:
- moves the description of intro patterns from under the intros tactic to its own subsection under "Tactics/Invocation of tactics"
- rewrites the material almost completely to make it simpler and accessible to the non-expert.  Description progresses from the most commonly to less commonly used parts.
- adds a grammar for intro patterns
- adds more (and simpler) examples, grouped at the end of the section

jfehrle commented on .gitattributes:
```diff
@@ -54,5 +54,3 @@ tools/CoqMakefile.in whitespace=blank-at-eol
 # CR is desired for these Windows files.
 *.bat whitespace=cr-at-eol,blank-at-eol,tab-in-indent
 
-* eol=lf
```
ignore this, it will go away

jfehrle commented:
I need help/feedback particularly on the following:
- Comments in the body of the rst file marked with "**...**" (boldface in the outline).  Most of these are separate lines/paragraphs
- Review comments I will make shortly
- Careful review that I've not dropped essential information in the process of rewriting.  I didn't fully understand parts of the existing documentation.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+                           : [= `intropatterns` ]
+   intropatterns           : `nonsimple_intropattern` ...
+                           : <empty>
+   ne_intropatterns        : `nonsimple_intropattern` ...
```
"ne" like non-empty? This whole business of the "empty" intro-pattern seems puzzling to me.

jfehrle replied:
Probably related to this odd-looking code.  Gotta have the `ne_intropatterns` for the parser.  Also seems like `eintros` should be usable without intropatterns.

```
       IDENT "intros"; pl = ne_intropatterns ->
          { TacAtom (CAst.make ~loc @@ TacIntroPattern (false,pl)) }
      | IDENT "intros" ->
          { TacAtom (CAst.make ~loc @@ TacIntroPattern (false,[CAst.make ~loc @@IntroForthcoming false])) }
      | IDENT "eintros"; pl = ne_intropatterns ->
          { TacAtom (CAst.make ~loc @@ TacIntroPattern (true,pl)) }
```

Zimmi48 replied:
I don't know if it's possible to simplify the code of the parser here, but at least the grammar in the manual should probably not mention this all empty / non-empty business.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+   nonsimple_intropattern  : `simple_intropattern`
+                           : *
+                           : **
+   naming_intropattern     : `pattern_ident`
```
I would also prefer if the `pattern_ident` indirection were removed. It is only making the grammar longer with no benefit (of course, once again this is better if it can be removed in the code as well).

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+* :n:`@ident` - use the specified name
+* :n:`?` - let Coq choose a name
+* :n:`? @ident` - generate a name that begins with :n:`@ident`
```
```suggestion
* :n:`?@ident` - generate a name that begins with :n:`@ident`
```

There should be no space, otherwise it will be parsed as two intropatterns. I suppose the grammar should be fixed as well to reflect that.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+The most common splitting patterns are:
+
+* split a hypothesis in the form :g:`A /\ B` into two
+  hypotheses :g:`H1: A` and :g:`H2: B` using the pattern :g:`[H1 H2]`
```
I usually recommend using `(H1, H2)` instead in this case...

jfehrle replied:
Why not always `(H1 & H2)`?

Zimmi48 replied:
Why not indeed? Although, if you have a left-associative operator (quite infrequent I suppose), you might need to use the `((H1, H2), H3)` which might still be preferred to `[[H1 H2] H3]`...

herbelin replied:
In the category of left-associative conjunctive operators, we missed the opportunity to make `prod` right-associative... (Too late, unless using a translator.)

Zimmi48 replied:
This can be done in stdlib2 though.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+  subgoals using the pattern :g:`[H1|H2]`.  The first subgoal will have the hypothesis
+  :g:`H1: A` and the second subgoal will have the hypothesis :g:`H2: B`.
+
+**:ref:`Example <_intropattern_conj_ex>` doesn't work for linking to the example**
```
Did you try removing the leading underscore?

jfehrle replied:
That works, thanks.  (But too much arcana in rst, I think.)

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+the label must precede a section header)"**
+
+Splitting patterns can have more than 2 names, for example :g:`[H1 H2 H3]`
+matches :g:`A /\ B /\ C`.  Patterns can be nested:
```
That's incorrect. Because of the implicit parenthesizing the actual pattern used to destruct this is `[H1 [H2 H3]]`, alternatively `(H1, (H2, H3))` or the much nicer (and thus recommended) `(H1 & H2 & H3)` that is made just for that purpose.

jfehrle replied:
The & form is recommended over the , form?  Both the [ ] and the , form are not recommended (maybe deprecated someday?)

Zimmi48 replied:
The point of the form with the comma is to look like a pair when destructuring something that can be understood as a pair. However; the intro-pattern with square brackets is more generic. The following cannot be done with any other intro pattern:

```
Inductive foo := | bla : nat -> nat -> foo | bar : nat -> foo.
Goal foo -> True.
intros [n1 n2|n3].
```

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+:g:`[[Ha|Hb] H]` can be used to split :g:`(A \/ B) /\ C`.
+
+Note that there is no equivalent to intro patterns for goals.  For a goal :g:`A /\ B`,
+use the :tacn:`split` tactic to replace the current goal with subgoals :g:`A` and :g:`B`.
```
Indeed, even if we had the idea: see #410.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+Note that there is no equivalent to intro patterns for goals.  For a goal :g:`A /\ B`,
+use the :tacn:`split` tactic to replace the current goal with subgoals :g:`A` and :g:`B`.
+For a goal :g:`A \/ B`, use :g:`apply or_introl` to replace the current goal with :g:`A`, or
```
```suggestion
For a goal :g:`A \/ B`, use :tacn:`left` to replace the current goal with :g:`A`, or
```

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+Note that there is no equivalent to intro patterns for goals.  For a goal :g:`A /\ B`,
+use the :tacn:`split` tactic to replace the current goal with subgoals :g:`A` and :g:`B`.
+For a goal :g:`A \/ B`, use :g:`apply or_introl` to replace the current goal with :g:`A`, or
+:g:`apply or_intror` to replace the current goal with :g:`B`.
```
```suggestion
:tacn:`right` to replace the current goal with :g:`B`.
```

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+For a goal :g:`A \/ B`, use :g:`apply or_introl` to replace the current goal with :g:`A`, or
+:g:`apply or_intror` to replace the current goal with :g:`B`.
+
+The other splitting patterns described below are not frequently used.
```
On the contrary, as I said before, the once with commas and `&` are very much recommended.

Zimmi48 replied:
Indeed, but the code of the standard library is not a good example of how people would write things today because it is mostly ancient code that has very rarely been updated.

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+                              : `equality_intropattern`
+                              : _
+                              : `naming_intropattern`
+   or_and_intropattern     : [ `intropatterns` | ... | `intropatterns` ]
```
I discovered that you don't have to have '|' to begin the 2nd..nth productions for a nonterminal.  I think we should remove them throughout the documentation because they are visual clutter, they make it harder to read and can confuse.

![image](https://user-images.githubusercontent.com/1253341/50367194-63f54480-0533-11e9-8028-679d3cc2a688.png)

Zimmi48 replied:
Sounds good.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+These patterns can be used when the hypothesis is an equality:
+
+* :n:`->` - replaces the right-hand side of the hypothesis with the left-hand
```
I just noticed that you sometimes put "s" at the end of this verbs and sometimes not.

jfehrle replied:
That's standard usage, right?  "[It] replaces ..." for a singular subject and "to replace" for the infinitive.  I didn't see a misuse.

Zimmi48 replied:
For instance:

```
* :n:`? @ident` - generate a name that begins with :n:`@ident`
* :n:`_` - discards the matched part (unless it is required for another
  hypothesis)
```

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+  constructor that itself has at least `n` arguments.  It performs a case
+  analysis over the hypothesis, as :n:`destruct` would, and applies the
+  patterns to the arguments of the constructor of :g:`I`.  Note that
+  that :n:`({+ p})` is equivalent to :n:`[{+ p}]`);
```
```suggestion
  that :n:`({+, @simple_intropattern})` is equivalent to :n:`[{+ @simple_intropattern}]`);
```

Zimmi48 replied:
I suggest we stick to BNF only in the grammars themselves and inside, `:n:` blocks, we use these "diagrams" but with actual tokens when we can.

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+                           : ?
+                           : `ident`
+   pattern_ident           : ? `ident`
+
```
I thought about whether a) to show all these productions together, b) to show a few productions at a time just before the text describing their semantics and c) both a and b.  The grammar in its current form here would be confusing for option b or c.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+**need example for & ... &**
+
+**A /\\ (exists x, B /\\ C /\\ D) is not valid for (a & x & b & c & d)**
```
How did you tested it? This works:

```coq
Parameters A B C D : Prop.
Goal A /\ (exists x : nat, B /\ C /\ D) -> True.
intros (a & x & b & c & d).
```

Zimmi48 replied:
There's no difference but you forgot to introduce the parameters in your context first, i.e. it should have been:

```
intros A x B C D (a & x & b & c & d).
```

jfehrle replied:
OK.  I didn't expect to see the `_` in this example.  The example would be simpler without that detail, but maybe that's the way it has to be.

```
Goal forall (A: Prop) (B: Prop) (C: Prop), A /\ (exists x:nat, B /\ C) -> True.
intros *.

1 subgoal
A, B, C : Prop
______________________________________(1/1)
A /\ (exists _ : nat, B /\ C) -> True

```

gives

Zimmi48 replied:
Here it is, tweaked so that it does not display any `_`:

```
Goal forall (A: Prop) (B: nat -> Prop) (C: Prop), A /\ (exists x:nat, B x /\ C) -> True.
intros *.
intros (a & x & b & c).
```

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+**A /\\ (exists x, B /\\ C /\\ D) is not valid for (a & x & b & c & d)**
+
+* :n:`*` - introduces one or more quantified variables from the result
```
... until there are no more quantified variables

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+* :n:`_` - discards the matched part (unless it is required for another
+  hypothesis)
+* if a disjunction pattern omits a name, such as :g:`[|H2]`, Coq will choose a name
+
```
I wanted to use boldface for, say ```:n:`?` ``` but ```**:n:`?`**``` doesn't interpret the `:n:`.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+* :n:`%term` - a pattern :n:`p{+ %term}` first applies :n:`{+ term}`
+  on the hypothesis to be introduced (as in :n:`apply {+, term}`) prior to the
+  application of the introduction pattern :n:`p`;
```
Note in passing. This would be nice to use actual tokens more systematically instead of `p`

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+**Sphinx says: "WARNING: undefined label: _intropattern_conj_ex (if the link has no caption
+the label must precede a section header)"**
+
```
I wanted to put the word "Example" in each of the two bullets so the user could find the example immediately rather than scrolling down several pages.  I suppose I could make each example a sub-sub-sub section to get around the problem.  People might even like to see that in the table of contents (though it will make it bigger).

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+  :n:`{+, p}` are used on the hypotheses generated by :tacn:`injection`; if the
+  number of patterns is smaller than the number of hypotheses generated, the
+  pattern :n:`?` is used to complete the list.
+
```
I didn't understand this, just copied the text with minor tweaks.

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+  constructor that itself has at least `n` arguments.  It performs a case
+  analysis over the hypothesis, as :n:`destruct` would, and applies the
+  patterns to the arguments of the constructor of :g:`I`.  Note that
+  that :n:`({+ p})` is equivalent to :n:`[{+ p}]`);
```
This is exactly equivalent to [A B C ...]?  If so, I would make it one sentence. The description is impenetrable.

herbelin replied:
The description is not up-to-date. It should be ``... constructor that itself has `n` arguments`` and it is not any longer equivalent. A correct formulation, hoping it is not too terse, could be ``Note that :n:`({+, p})` is equivalent to :n:`[{+ p}]` for a list made of as many patterns as the constructor has arguments`.

Note that in this case, there is a tolerance. The count of arguments is done both including or excluding let-ins and both cases are allowed. I.e., in:
```
Inductive I := C : forall a:nat, let b:=a in forall c:nat, I.
Goal forall x:I, x = x.
```
both `intros (a,b,c)` and `intros (a,c)` are allowed, and the latter is the same as `intros (a,?,c)`.
```

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+  right-associative binary inductive constructors such as :g:`conj` or
+  :g:`ex_intro`; for instance, a hypothesis with type
+  :g:`A /\(exists x, B /\ C /\ D)` can be introduced via pattern
+  :n:`(a & x & b & c & d)`;
```
Couldn't get the example to work.  A generalization of [A B C]?  Didn't fully understand.

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+   is a disjunctive or conjunctive pattern
+   :n:`[{+| @intro_pattern_list}]`, the completion of :n:`@intro_pattern_list`
+   so that all the arguments of the i-th constructors of the corresponding
+   inductive type are introduced can be controlled with the following option:
```
Couldn't tell if this is important or not, or if it could be said more simply.

Zimmi48 replied:
I just noted that this option, as its name suggests only impacts the last intro pattern. In the other cases, it has no effect and does not make `[]` behave like in SSReflect. I'm wondering what was the objective in introducing this option and whether we should remove it (simpler language, less options, less confusion).

herbelin replied:
The flag `Bracketing Last Introduction Pattern` is indeed a compatibility flag. It would probably be less confusion to eventually deprecate it.

Note that ssreflect introduction patterns extended an original mechanism of introduction patterns (in Coq 6.2) which was "procedural". For better compositionality, the introduction patterns evolved in vanilla Coq into a "structured" notion of introduction patterns but ssreflect kept the initial implementation of Coq 6.2. When the pattern is complete, in the sense of giving an introduction pattern for each of the argument of the constructor, there is no difference. There is a difference when it is incomplete. The original 6.2 implementation is such that, e.g., `[ | ] H` is dependent on the inductive structure of the first hypothesis. For instance, if the goal is `forall n:nat, n = 0 -> 0 = n`, the old implementation made that `[ | ] H` was introducing `H : n=0` in the first resulting goal but `H : nat` (the predecessor) in the second goal. On the contrary, the current (vanilla) design has the property that a list `p1 ... pn` of simple introduction patterns applies pointwise to the next `n` first quantified hypotheses or let-ins of the goal.

I don't know what is the best way to explain the question of completion. Maybe can it be said that if `[pat1 ... patn]` has less simple patterns than the number of arguments in the constructor, it automatically expands to `[pat1 ... patn ? ... ?]` with as many extra introduction patterns so as to match the arity of the constructor. Note that let-ins are included in the count.

Zimmi48 replied:
Yes, I think the explanation you give in the last paragraph is a good one.
@jfehrle Could you deprecate the `Bracketing Last Introduction Pattern` as part of this PR?

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+.. _intropattern_conj_ex:
+
+   .. example:: intro pattern for /\\
```
The text on the right appears in the header for the example in the output.  Seems useful

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+multiple hypotheses or subgoals.  Common tactics that accept intro patterns
+include :tacn:`intros`, :tacn:`destruct` and :tacn:`apply`.
+
+.. productionlist::        coq
```
"The argument to productionlist serves to distinguish different sets of production lists that belong to different grammars."

Zimmi48 replied:
Interesting, it's not how it was used so far in the manual...

jfehrle replied:
I expect that needs untangling at some point.  `productionlist:: coq` is used 9 time in the .rst files.

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+                           : <-
+                           : [= `intropatterns` ]
+   intropatterns           : `nonsimple_intropattern` ...
+                           : <empty>
```
It represents the empty production.  Otherwise it would just be an empty line, which could be interpreted as just a blank line.

![image](https://user-images.githubusercontent.com/1253341/50367732-0d8a0500-0537-11e9-90b7-3683c664df9c.png)

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -82,7 +82,7 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
                    : `term` ... `term`
 
-+ In a bindings list of the form :n:`{* (ref:= term)}`, :n:`ref` is either an
++ In a bindings list of the form :n:`{+ (ref:= term)}`, :n:`ref` is either an
```
I think it's OK.  "bindings list" can only mean a list of bindings.  "binding list" could mean "list of bindings" or "binding" could be taken as a modifier of "list" (like "binding agreement" or "sorted list").  The phrase appears in 9 places in the file, BTW.

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+.. productionlist::        coq
+   simple_intropattern     : `simple_intropattern_closed`
+                           : % `operconstr` ... % `operconstr`
```
It's defined as: ```LIST0 ["%"; c = operconstr LEVEL "0" -> { c } ]"``` in simple_intropattern.

Zimmi48 replied:
BTW you should probably change this `operconstr` into `term` as it seems we rarely distinguish between the various term representations in the grammars in the manual.

jfehrle replied:
I'd like to understand that better first.  "term" is not even a non-terminal in the grammar.

Zimmi48 replied:
The internal name for "term" is "constr" in Coq for historical reason (when Coq was called "The Construction Engine" from the "Calculus of Constructions"). Developers have wondered whether it should be renamed to "term" but it is used too much for such a change to be worthwhile... However, in the manual we use "term" instead to be less confusing.

herbelin replied:
The PR is still evolving, so maybe this is not yet the right time to comment. A few remarks though.

Yes, `operconstr` is the internal parser name of the recursive grammar of terms. This grammar has two useful entry points: `constr` which is `operconstr` at level 8 (to be used in positions where applications needs to be parenthesized to avoid collisions with the rest of the grammar), and `lconstr` which is `operconstr` at level 200 (to be used in a context where the expression is already surrounded by delimiters).

We could in principle be explicit in the reference manual on the different uses of level, but it has been considered simple to call `term` any of `constr`, `lconstr` and `operconstr`. Some improvements could be done though.

jfehrle commented:
> A remark in passing: I think there should be a note saying that intro-patterns are an advanced feature that allows to write proofs more quickly but that beginners do not need (everything can be done by applying simple tactics, like intro, destruct and rewrite).

I can add a note, but destruct itself uses intro patterns.  destruct can do the equivalent of [A [B|C]]?  And the note should probably mention how to use simple tactics in place of intro patterns.

anton-trunov commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+  itself is not a pattern for on-the-fly application of lemmas (note:
+  syntax is in experimental stage)  **huh?**
+
+**need example, maybe simplify "intros [= -> %IHl]." from List.v?**
```
Could this [example](https://github.com/tchajed/coq-tricks/blob/master/src/IntroPatterns.v) be of any help here?

jfehrle replied:
Thanks, I'll adapt that.  But without some explanation, understanding it takes a couple minutes of experimentation:
- `intros` would just add `H : S (length ys) = 1`
- `intros [=]` would reduce (right term?) to add `H0 : length ys = 0`
- `intros [=->%length_zero_iff_nil]` applies the resulting equality `l=[]`

I wish Coq had a way to show these multiple substeps for this example (and many, many others).  It's not so easy to understand what Coq is doing when there are multiple substeps.  I think this would be very helpful for learners.  But creating such a feature might be difficult but maybe feasible.  Certainly I'd have to learn far more about tactic internals.  (What do you think?)

It may be helpful to have a way to do diffs that comparing the result of applying 2 alternative tactics.  (Maybe).

Off topic: have you tried the diffs feature in master?  It will be in 8.9, but it looks like 2 needed bug fix PRs for it may not make it into 8.9 :-(.

Zimmi48 replied:
The `Info` command was an attempt to display such a decomposition but it is unfinished.

anton-trunov replied:
> I wish Coq had a way to show these multiple substeps for this example (and many, many others).

I fully agree. Although, I'd say that stepping through a proof (with all those `;`) without dismantling it first would be a the most desirable feature.

> Off topic: have you tried the diffs feature in master?

Not yet. Does ProofGeneral support it?

jfehrle replied:
@anton-trunov No, diffs are not in PG yet.  I was hoping someone who's worked on PG would volunteer to do the small amount of work needed for diffs in PG once 8.9 is out.  Perhaps unrealistic, I'll see.

> I'd say that stepping through a proof (with all those ;) without dismantling it first would be a the most desirable feature.

Maybe it would not be hard to add an option that would treat ; as if it was "." for non-LTAC code.   I suspect it may be complex to robustly extend the concept to LTAC, which I don't know well.   Would an option that doesn't work in LTAC still be worthwhile?

Zimmi48 replied:
You cannot simply replace `;` with `.` and hope for the code to work though unless you add the proper goal selectors.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+Intro patterns let you specify the name to assign to variables and hypotheses
+introduced by tactics.  They also let you split an introduced hypothesis into
+multiple hypotheses or subgoals.  Common tactics that accept intro patterns
+include :tacn:`intros`, :tacn:`destruct` and :tacn:`apply`.
```
Yes, but I think we should have more cross-links between the Tactics chapter and the SSR chapter.

jfehrle replied:
I think such links should be considered as a group and done consistently after the tactics and SSR chapters have been extensively revised.  I don't want to consider that now.

Zimmi48 commented:
>I can add a note, but destruct itself uses intro patterns. destruct can do the equivalent of [A [B|C]]? And the note should probably mention how to use simple tactics in place of intro patterns.

Indeed, you are right that when using `destruct`, you need to use intro patterns if you want to avoid auto-generated names. In the following context:

```
Goal (True /\ (True \/ True)) -> True.
```
these are equivalent:
```
intros [A [B | C]].
```
```
intro H. destruct H as [A [B | C]].
```
```
intro H. destruct H as [A H]. destruct H as [B | C].
```

jfehrle commented:
> A remark in passing: I think there should be a note saying that intro-patterns are an advanced feature that allows to write proofs more quickly but that beginners do not need (everything can be done by applying simple tactics, like intro, destruct and rewrite).

After consideration, I think such a note doesn't really fit the section.  Basic use of the feature is simple to understand and I believe should be used routinely (e.g. give a name in `intros`).  Saying it is advanced will needlessly scare people away.  Why would a user assume they **must** use any particular feature they see in the manual?  The text here already identifies the most common uses.

Suitably adapted, the note could be part of an overview section that gets into what tactics are primitives vs composites/conveniences.

jfehrle commented:
I made many changes, please take another look.  Still several points I don't understand completely:

For patterns, is this right?
- (a, b) - recommended for matching A /\ B instead of (a & b) - why not (a & b)?
- (a & b & c) - recommended for matching A /\ B /\ C,  This matches things that are nested right associatively.  You need ((a & b) & c) to match (A /\ B) /\ C.  This will work for /\ and exists and what else?.
- (HA, (HB, HC)) required to match A /\ B /\ C (have to match the structure)
- [HA HB] - no longer recommended (less flexible).  [HA [HB HC]] required to match A /\ B /\ C
- [HA|HB] - only way to match A \\/ B

What would match (HA, HB, HC)?
Is the comma form the same as [HA HB] except for the syntax?

jfehrle commented:
![image](https://user-images.githubusercontent.com/1253341/50390716-8e8bfc80-06ef-11e9-8115-8f939ead6084.png)

Seems like the expression (syntax tree) on the second line should be more like (constr1 a1 (constr2 a2 (constr3 a3 a4))).

Zimmi48 commented:
- **`& ` intro-pattern**
  >This will work for /\ and exists and what else?.

  Any inductive type with a single constructor with two parameters.

- >What would match (HA, HB, HC)?
  >Is the comma form the same as [HA HB] except for the syntax?

  Yes, so for instance:

  ```
  Inductive and3 (A B C : Prop) : Prop := conj3 : A -> B -> C -> and3 A B C.
  Goal and3 True True False -> False.
  intros (_, _, ?). assumption.
  Qed.
  ```

  Obviously `intros [_ _ ?]` would have worked as well. However, `intros (_ & _ & ?)` would not because this implies nested two-parameter constructors.

- >Seems like the expression (syntax tree) on the second line should be more like `(constr1 a1 (constr2 a2 (constr3 a3 a4)))`.

  I think the point was to show an equivalent intro pattern, not a term being matched.

Zimmi48 commented on doc/sphinx/language/coq-library.rst:
```diff
@@ -105,17 +105,17 @@ subclass :token:`form` of the syntactic class :token:`term`. The syntax of
-        : | exists2 `ident` [: `specif`], `form` & `form`          (ex2)
-        : | `term` = `term`                                    (eq)
-        : | `term` = `term` :> `specif`                          (eq)
+        : False                                          (False)
```
There is now a misalignement that would be fixed by removing the additional spacing before `True` on the first line. This is the only place where I've noticed this.

Zimmi48 commented on doc/sphinx/language/gallina-specification-language.rst:
```diff
@@ -524,33 +524,33 @@ The Vernacular
-                      : | Proof . … Defined .
-                      : | Proof . … Admitted .
+                      : Proof . … Defined .
+                      : Proof . … Admitted .
```
You have missed some leading `|` just below in the attributes grammar.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -33,10 +33,13 @@ extends the folklore notion of tactical) to combine those atomic
 
 Invocation of tactics
--------------------------
+~~~~~~~~~~~~~~~~~~~~~
```
You have missed a leading `|` just below in the "tactic invocation" grammar.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,277 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+   ne_intropatterns        : `intropattern` ... `intropattern`
+   intropatterns           : `intropattern` ... `intropattern`
+                           : `empty`
+   empty                   :
```
I like a lot this way you found to make the empty case explicit, and it is also clear why this is needed. However, I would avoid repeating the ``` `intropattern` ... `intropattern` ``` part. So I suggest changing the above to:

```
    intropatterns           : `ne_intropatterns`
                            : `empty`
    ne_intropatterns        : `intropattern` ... `intropattern`
    empty                   :

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,277 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+                           : ?
+                           : ?`ident`
+   or_and_intropattern     : [ `intropatterns` | ... | `intropatterns` ]
+                           : ()
```
I don't know this intro pattern and it doesn't look like it is explained below.

Zimmi48 replied:
Ping: this intro-pattern should be removed from the grammar.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,277 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+* :n:`?@ident` - generate a name that begins with :n:`@ident`
+* :n:`_` - discards the matched part (unless it is required for another
+  hypothesis)
+* if a disjunction pattern omits a name, such as :g:`[|H2]`, Coq will choose a name
```
A conjunctive or disjunctive pattern. In fact the simplest (and most useful) example is probably the common `[]`.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,277 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+  :g:`ex_intro`.
+  :ref:`Example <intropattern_ampersand_ex>`
+
+The other splitting patterns described below are not frequently used.
```
In the current form, I still disagree with this sentence. For instance, I do use *very* frequently the `->` / `<-` and `*` intro patterns.

jfehrle commented:
![image](https://user-images.githubusercontent.com/1253341/50390815-b0d24a00-06f0-11e9-854d-08ff88b497f0.png)

What's an example with more than 2 arguments?

jfehrle commented on doc/sphinx/addendum/implicit-coercions.rst:
```diff
@@ -186,7 +186,7 @@ Figure :ref:`vernacular` as follows:
    assumption : assumption_keyword assums .
    assums : simple_assums
-          : | (simple_assums) ... (simple_assums)
+          : (simple_assums) ... (simple_assums)
```
Looks like there are also backticks where they don't belong, e.g. `fail` in generralized-rewriting.mlg:

```
.. productionlist:: rewriting
   strategy : "(" `s` ")"
   lemma : `c`
   lemma_right_to_left : "<-" `c`
   failure : `fail`
     :
```

jfehrle replied:
Also, the quotes in `"("` appear in the output, I think they should be removed.

jfehrle replied:
The layout for that section is a bit strange.  addendum/generalized-rewriting.html#strategies-for-rewriting.

![image](https://user-images.githubusercontent.com/1253341/50411805-f4dc5200-07b7-11e9-89c2-1ddccf74d291.png)

![image](https://user-images.githubusercontent.com/1253341/50411813-01f94100-07b8-11e9-990d-204af95ec9e2.png)

But I don't want to get sidetracked making non-trivial fixes to all the grammars.

Zimmi48 replied:
Indeed, only nonterminals.

herbelin commented:
> Browsing through the tactics using intro patterns in the chapter, I notice that:
> - `remember` and `eremember` should probably use `naming_intropattern` instead of `ident` for the `eqn:` clause; `destruct` uses `naming_intro_pattern` for its `eqn:` clause when it should be `naming_intropattern`;
> - the `assert`, `eassert`, `specialize`, `injection`, `einjection`, (`dependent` / `simple`) `inversion`, (`dependent`) `inversion_clear` tactics should use the correct grammar token (probably `simple_intropattern`) instead of the current fake `intro_pattern` in their `as` clause;
> - `destruct` uses `disj_conj_intro_pattern` but this token has been renamed (same for `induction` and `functional induction`)

For `injection`, there is a little syntactic discrepancy. It would be more natural to move the syntax from `injection ... as ipat1 ... ipatn` to `injection ... as [= ipat1 ... ipatn]`.

There are otherwise also inconsistencies of terminology in the code (e.g. we use `Tacarg.wit_intro_pattern` for denoting the grammar entry `simple_intropattern` - end of `pltac.ml`).

Zimmi48 commented:
>Will it be possible to make Info return an always-accurate list of tactics?

I don't think so. Honestly, I wouldn't try fixing this `Info` tactic.

Zimmi48 commented:
I did not look at the code again, only replied to comments. @jfehrle, let me know when you think this is ready to undergo a final review (or if you have more questions).

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,277 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+  the on-the-fly application of lemmas: :n:`p{+ % @term}` where :n:`p`
+  itself is not a pattern for on-the-fly application of lemmas (note:
+  syntax is in experimental stage)  **huh?**
```
Will do.  FWIW: Better to say "Anyway" here.  I don't recall ever using "Anyways", even in casual speech.

```
“Anyways” with the S is considered slang, and is a part of nonstandard, colloquial, or
informal English. Furthermore, since “anyway” is an adverb and it is impossible for
adverbs to be plural. In newspapers or in any example of formal writing, you will
almost never see the word “anyways” used.
```

Zimmi48 replied:
Interesting, I had read somewhere else that the two were equivalent...

jfehrle commented:
> Browsing through the tactics using intro patterns in the chapter, I notice that:
> - `destruct` uses `disj_conj_intro_pattern` but this token has been renamed (same for `induction` and `functional induction`)

Not so simple, what's allowed for `destruct` is an `or_and_intropattern ` or an `ident`, which is a subset of `simple_intropattern`.  There are 11 uses of disj_conj_intro_pattern in Sphinx, and in the .mlg's 6 uses of induction_clause_list and 6 uses of as_or_and_ipat.  Does it make sense for someone to change `destruct` and the other referencing tactics to use `simple_intropattern` in the name of consistency/simplicity?  @herbelin, any thoughts?

```
simple_tactic:
      | IDENT "destruct"; icl = induction_clause_list ->

inclusion_clause_list:
    [ [ ic = LIST1 induction_clause SEP ","; el = OPT eliminator;
        cl_tolerance = opt_clause ->

  induction_clause:
    [ [ c = destruction_arg; pat = as_or_and_ipat; eq = eqn_ipat;
        cl = opt_clause ->

  as_or_and_ipat:
    [ [ "as"; ipat = or_and_intropattern_loc -> { Some ipat }
      | -> { None } ] ]

  or_and_intropattern_loc:
    [ [ ipat = or_and_intropattern -> { ArgArg (CAst.make ~loc ipat) }
      | locid = identref -> { ArgVar locid } ] ]

  identref:
    [ [ id = ident -> { CAst.make ~loc id } ] ]
```

jfehrle commented:
> Browsing through the tactics using intro patterns in the chapter, I notice that:
> - the `assert`, `eassert`, `specialize`, `injection`, `einjection`, (`dependent` / `simple`) `inversion`, (`dependent`) `inversion_clear` tactics should use the correct grammar token (probably `simple_intropattern`) instead of the current fake `intro_pattern` in their `as` clause;

For each of these, would it make sense to change the grammar to use more general nonterminals (simple_intropattern, etc.) for consistency?
- `specialize` uses `intropattern`.
- `injection` and 'einjection` use `intropattern_list`, a definition I can't find in the .mlg files.
- `inversion` etc. uses `and_or_and_ipat`.

jfehrle commented:
For this one, why does the `destruct` result go into the goal?
```
Inductive and3 (A B C : Prop) : Prop := conj3 : A -> B -> C -> and3 A B C.
Goal (and3 True True False -> False) -> False.
intros. destruct H as [].
```

Whereas this `destruct` leaves things in the hypotheses:

```
Goal and3 True True False -> False.
intros A. destruct A as [a b c]. 
```

Makes me think my description is at best incomplete.

herbelin commented:
> ```coq
>  Inductive and3 (A B C : Prop) : Prop := conj3 : A -> B -> C -> and3 A B C.
> Goal (and3 True True False -> False) -> False.
> intros. destruct H as [].
> ```

Here, you `destruct` an hypothesis in type `False` with a side condition of type `and3 True True False`. The type `False` leaves no subgoal and what remains is only the side condition.

> `injection` and `einjection` use `intropattern_list`, a definition I can't find in the `.mlg` files.

There is a grammar of entry names for use in `TACTIC EXTEND` and `Tactic Notation` which is a level of indirection on top of the more primitive `GRAMMAR EXTEND`. In `TACTIC EXTEND` and `Tactic Notation`, you can add the suffix `_list` or `_opt`, as well as the pair of prefix/suffix `ne_`/`_list` to build a new entry from the existing one. This is the same as using `LIST0`, `OPT`, `LIST1` respectively in the context of `GRAMMAR EXTEND`.

To make things more complicated, such `intropattern` in `TACTIC EXTEND` and `Tactic Notation` is actually bound in `GRAMMAR EXTEND` to `simple_intropattern` (via an expansion by `coqpp` to `wit_intropattern` which itself is bound to the `GRAMMAR EXTEND` entry `simple_intropattern` in `pltac.ml`). So, all the cited tactics actually use `simple_intropattern` according to `GRAMMAR EXTEND` terminology. Similarly, tactics using `intropattern_list` in `TACTIC EXTEND` and `Tactic Notation`, like `injection` does, are actually using `LIST0 simple_intropattern` when seen from the point of view of `GRAMMAR EXTEND`.

So, to simplify the terminology, we may avoid this kind of indirection. There are basically two options:
- We keep `simple_intropattern` as it is and force users to rename all uses of `intropattern` (thus also of `intropattern_opt`, `intropattern_list`, `ne_intropattern_list`) in `TACTIC EXTEND` and `Tactic Notation`. Then, in a latter step, we may optionally reintroduce the word `intropattern` to mean the current `nonsimple_intropattern` (as was proposed above in the `GRAMMAR EXTEND` context).
- We rename `simple_intropattern` into `intropattern` in `GRAMMAR EXTEND` so that it is consistent with what users use in `TACTIC EXTEND` and `Tactic Notation`. Optionally, we may also make public the name `nonsimple_intropattern` so that it can be used in `TACTIC EXTEND` and `Tactic Notation`.

> Does it make sense for someone to change `destruct` and the other referencing tactics to use `simple_intropattern` in the name of consistency/simplicity? @herbelin, any thoughts?

As said in [#9288](https://github.com/coq/coq/pull/9288#issuecomment-450347616), some people tends to see the word destruct as a more general concept than just splitting an inductive object into its component. However, the grammar of `destruct` is already complex, including `eqn:`, `in`,  multiple arguments. Changing it to support a more general concept of destruction which would include any kind of `simple_intropattern` action (i.e. not only `[... | ... | ...]`, but also `->`, `[=]`, `%constr`, `_`, `?id`, ...) seems a bit too risky change.

jfehrle commented:
> There is a grammar of entry names for use in `TACTIC EXTEND` and `Tactic Notation` which is a level of indirection on top of the more primitive `GRAMMAR EXTEND`.

OK.  Is there developer documentation for the coqpp extensions in .mlg files such as `GRAMMAR EXTEND`?  If not, this seems so basic that I think it should be created sometime soon.  And some explanation of how `register_grammar` etc. works--I don't think I've seen that in, say, the Camlp5 documentation.  @Zimmi48, is there someone who might be willing to create that?

For the simple_intropattern/intropattern inconsistency, "force users to rename all uses of `intropattern`" means developers of plugins not in the repository would be affected.  That sounds undesirable.  "We rename simple_intropattern into intropattern in GRAMMAR EXTEND" sounds like it wouldn't impact others.  Is this just a mechanical change I could make (likely as a separate PR)?

What does "wit_" stand for?  "with"?  "witness"??

Also, I notice the doc doesn't describe the use of "as <intropattern>" for case, ecase,  Add that?

`intros` works with no arguments--shouldn't `eintros` also work with no arguments?  That would let me drop ne_intropatterns from the grammar in the doc.  And eintros is not documented.

jfehrle commented:
> We rename simple_intropattern into intropattern 

It just occurred to me that there already is an `intropattern` defined in ltac/g_tactic.mlg.  That would need to be renamed, right?  There are only 3 uses.  On the other hand, it's a good name for what it does

```
  intropattern:
    [ [ l = simple_intropattern -> { l }
      | "*"  -> { CAst.make ~loc @@ IntroForthcoming true }
      | "**" -> { CAst.make ~loc @@ IntroForthcoming false } ] ]
```

![image](https://user-images.githubusercontent.com/1253341/50532029-ed72bc80-0ac7-11e9-8e51-93179b3714c7.png)

jfehrle commented:
Re: Bracketing Last Introduction Pattern

It's hard to make enough sense of this to write it up.

```
Goal True /\ False -> False.
intros []. (* H : True, H0: False |- False *)
Undo.
Unset Bracketing Last Introduction Pattern.
intros []. (* |- True -> False -> False *)
```

I think the second part of your example (in which True /\ False is transmogrified into True -> False) would be too confusing to the reader.  At the least it would require deeper explanation that would be out of place here.

The use of "completion" in the current text is confusing.  If there aren't enough names provided to match the term, then Coq will generate names.  And if the option is off, the parts of the pattern that aren't named are more-or-less left untouched.  (How can I express this in the doc?)

How about something like this?

```
.. flag:: Bracketing Last Introduction Pattern

   Controls how to handle the last introduction pattern when it is a
   disjunctive or conjunctive pattern that doesn't give enough names to
   fully match the term.  If set (the default), |Coq| generates additional names
   to complete the splitting.  Unsetting the option will _____________, behavior that is more
   similar to |SSR|'s ``intro`` patterns.
```

I'm also wondering whether a) this is really important enough to include and b) whether some of this is quirky/strange behavior that we should perhaps look at changing because it is confusing/hard to describe.  It doesn't make sense to document odd corner cases unless they're really useful--otherwise you end up spending a lot of time on unimportant stuff.

jfehrle commented:
I made many updates, please take a look, particularly at the splitting patterns.  There are also inline comments marked in bold (`**`) -- please respond to them.  I'm not entirely satisfied with the result, but I think it is a big step forward.  I'm eager to get this done.

@herbelin For "intropattern" and GRAMMAR EXTEND, can I have both "simple_intropattern" and "intropattern" refer to "simple_intropattern" from the main grammar?  That would not break anyone's code.  Is there any way to deprecate the use of "intropattern" in the grammar.  (Adding to the confusion, I realized that I added "intropattern" to the main grammar which maybe I should reverse.)

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -71,22 +74,22 @@ specified, the default selector is used.
                    : `term` ... `term`
 
-+ In a bindings list of the form :n:`{* (ref:= term)}`, :n:`ref` is either an
++ In a bindings list of the form :n:`{+ (@ref:= @term)}`, :n:`@ref` is either an
```
If `ref` was never defined in a grammar, it would be worth adding it to the grammar above.

jfehrle replied:
I added:

```
     ref           : ident
                   : INT
```

And I noticed that "INT" is a poor name, since it only includes natural numbers.  Maybe I'll clean that up in another PR.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -33,20 +33,23 @@ extends the folklore notion of tactical) to combine those atomic
-                       : |tactic .
+  .. productionlist:: sentence
+     tactic_invocation : `toplevel_selector` : `tactic`.
+                       : `tactic` .
```
```suggestion
                       : `tactic`.
```

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -71,22 +74,22 @@ specified, the default selector is used.
+Tactics that take a term as an argument may also support a bindings list,
 so as to instantiate some parameters of the term by name or position.
 The general form of a term equipped with a bindings list is ``term with
 bindings_list`` where ``bindings_list`` may be of two different forms:
```
Should use ``` :n:`@term with @bindings_list` where :token:`bindings_lists`... ```

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -71,22 +74,22 @@ specified, the default selector is used.
++ In a bindings list of the form :n:`{+ (@ref:= @term)}`, :n:`@ref` is either an
   :n:`@ident` or a :n:`@num`. The references are determined according to the type of
-  ``term``. If :n:`ref` is an identifier, this identifier has to be bound in the
+  ``term``. If :n:`@ref` is an identifier, this identifier has to be bound in the
```
```suggestion
   :n:`@term`. If :n:`@ref` is an identifier, this identifier has to be bound in the
```

Same change to do three other times in this paragraph, and one missing `@` in the next paragraph.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+Intro patterns let you specify the name to assign to variables and hypotheses
+introduced by tactics.  They also let you split an introduced hypothesis into
+multiple hypotheses or subgoals.  Common tactics that accept intro patterns
+include :tacn:`apply`, :tacn:`intros` and :tacn:`destruct`.
```
It's not `apply` but `apply ... in` which is a very different (and much less known) variant which supports an `as` clause. Thus, I would not cite `apply` in this list. I would rather cite `assert` or `specialize`.

jfehrle replied:
I see your point.  My idea for revising doc for `apply` and other tactics is to show the full syntax covering all options first, then identifying the most common options, followed by the details, akin to this section.  So there may only be a single index entry for `apply`.  I'll put in `assert`.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+   equality_intropattern   : ->
+                           : <-
+                           : [= `intropatterns` ]
+   or_and_intropattern_loc : or_and_intropattern
```
```suggestion
   or_and_intropattern_loc : `or_and_intropattern`
```

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+* split a hypothesis in the form :n:`A /\ B` into two
+  hypotheses :g:`H1: A` and :g:`H2: B` using the pattern :g:`(H1 & H2)` or
+  :g:`(H1, H2)` or :g:`[H1 H2]`
```
```suggestion
  :g:`(H1, H2)` or :g:`[H1 H2]`.
```

jfehrle replied:
FWIW: Technically, this is a sentence fragment (lacking a subject), therefore it doesn't really need a period at the end, although in the second bullet point I put a period to separate the fragment from the sentence that follows up.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+  :ref:`Example <intropattern_disj_ex>`
+* split a hypothesis in either of the forms :g:`A /\ B` or :g:`A \/ B` using the pattern :g:`[]`.
+
+**A node has to be of type Prop to be matched/split/named, right?**
```
No.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+**A node has to be of type Prop to be matched/split/named, right?**
+
+**what other common operators are intro patterns useful for?  ->? <->? forall?  I would like to
+mention them**
```
Yes `<->`: use a conjonctive intro-pattern as with `/\`. `forall` and `->` obviously but because intro patterns are used to introduce variables / hypotheses in the context, that's all.

jfehrle replied:
Ah, I should have said "which splitting patterns are useful for other operators?".  <-> splits nicely but I suppose `forall` does not.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+* :n:`[ {+| @intropatterns} ]` - splits an inductive type that has
+  :ref:`multiple constructors <intropattern_cons_note>`
+  such as :n:`A \/ B`
+  into multiple subgoals.  The number of ``intropatterns`` must be the same as the number of
```
```suggestion
  into multiple subgoals.  The number of :token:`intropatterns` must be the same as the number of
```

jfehrle replied:
FWIW It's awkward to have both `intropattern` and `intropatterns` non-terminals--makes it hard to distinguish singular and plural references.

jfehrle replied:
I think we should banish plural non-terminal names like `intropatterns` in favor of `intropattern_list`.  But not in this PR.

Zimmi48 replied:
Well, I agree at least that as part of this PR, the token be named `introppatern_list` instead of `intropatterns`.

jfehrle replied:
Done, but only in the doc.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+  such as :n:`A \/ B`
+  into multiple subgoals.  The number of ``intropatterns`` must be the same as the number of
+  constructors for the matched part.
+* :n:`[ {* @intropattern} ]` - splits an inductive type that has a
```
I think you meant ``` :n:`[ @intropattern ]` ``` here. But I would say that you can simply remove this bullet point and change the `+` to `*` in the previous bullet point as the explanation above applies to this case as well (understanding "multiple" in the large sense) and anyways the `( , )` intro-pattern is recommended for this case.

jfehrle replied:
Ah, no.  `[]` works for both A /\ B and A \\/ B, whereas the 2 prior bullets are specific to one or the other.  I will add "(a special case of the last 2 bullets)".

(Where are the recommendations from?  And why does it matter to the user?  Both forms are valid and AFAIK we're not planning to remove one of them.]

Zimmi48 replied:
The disjunctive form is the most general because the conjunctive form is just its specialization to inductive types with a single constructor. This is what I meant.
OTOH, `[]` is very special.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+**it's split into multiple hyps when the inductive type is (A B C:Prop) A -> B -> C -> something
+but probably not for terms with other types, e.g. nat.  Is this behavior tied to the intro pattern or more
+tied to ->?**
```
I did not understand this question. Can you reformulate?

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+   .. example:: [=->%length_zero_iff_nil] intro pattern
+
+      .. from https://github.com/tchajed/coq-tricks/blob/master/src/IntroPatterns.v
```
I suggest you put the credits in the text of the example.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+      .. from https://github.com/tchajed/coq-tricks/blob/master/src/IntroPatterns.v
+
+      | :g:`Theorem length_zero_iff_nil (l : list A):`
+      |    :g:`length l = 0 <-> l=[].`
```
Use `.. coqdoc::`, or `Check` inside a `.. coqtop::` block

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+      | :g:`Theorem length_zero_iff_nil (l : list A):`
+      |    :g:`length l = 0 <-> l=[].`
+
+      * `intros` would add :g:`H : S (length ys) = 1`
```
The ordering is weird. It would be natural if this were commenting on the code above but this is commenting on the text below. What about reversing completely the order in this example: first the code, then the bullet points, then the theorem statement.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+      * `intros` would add :g:`H : S (length ys) = 1`
+      * `intros [=]` would additionally apply :tacn:`injection` to :g:`H` to yield :g:`H0 : length ys = 0`
+      * `intros [=->%length_zero_iff_nil]` applies the theorem, making H the equality :g:`l=[]`,
+        which is then applied as for :g:`->`.  (Note :g:`nil` is the same as :g:`[]`.)
```
Rather than adding this note, you could make the theorem statement talk about `nil` instead of using the notation.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+.. _intropattern_injection_ex:
+
+   .. example:: [=->%length_zero_iff_nil] intro pattern
```
I would call this example “Advanced combination illustrating `->`, `%` and `[=]`” and probably put it at the end.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+      .. coqtop:: reset none
+
+         Goal forall (A: Prop) (B: nat -> Prop) (C: Prop), A /\ (exists x:nat, B x /\ C) -> True.
```
I suggest putting the parameters in a `Variables` listing as in the previous example so that they don't pollute the context.

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+
+Patterns can be nested: :n:`[[Ha|Hb] H]` can be used to split :n:`(A \/ B) /\ C`.
+
+Note that there is no equivalent to intro patterns for goals.  For a goal :g:`A /\ B`,
```
Strictly speaking, a goal is the hypotheses plus the conclusion, right?  The same as "proof context"?  "proof context" or "context" seems clearer than "goal".  References such as "For a goal :g:`A /\ B`" on this line seem very natural.  Or do we want to be strict and always use the term"conclusion"?

Zimmi48 replied:
Hum. Strictly speaking, we may include the context in what we call "a goal" (and this is what `Show Goal` does). But when we speak more specifically of what appears below the bar, we generally say "goal" as well. Here, I meant the conclusion of the latter, i.e. what goes after the last `->`

jfehrle replied:
Huh?  Where is the `->` here?  "conclusion" seems out of place.  I'm talking about applying `split` in this case:

```
1 subgoal
A, B : Prop
______________________________________(1/1)
A /\ B
```
The paragraph is not about intro patterns per se; it's meant to provide clarifying contrast for when you have `A /\ B` below the line.  Perhaps it would be better if the first sentence was "Note that you must use other tactics to split up goals."

Or I could drop the paragraph entirely.

Zimmi48 replied:
The conclusion is the full goal because there is no `->`. Anyway, it seems that the "clarification" I proposed wasn't one in reality. I prefer that we keep your paragraph as is. I think it is really useful.

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+  with the right-hand side of the hypothesis.
+* :n:`[= {*, @intropattern} ]` - If the product is over an equality type,
+  applies either :tacn:`injection` or :tacn:`discriminate`
+  instead of :tacn:`destruct`; if :tacn:`injection` is applicable, the intropatterns
```
Good point, leftover from the old text.

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+   conjunctive pattern that doesn't give enough simple patterns to match
+   all the arguments in the constructor.  If set (the default), |Coq| generates
+   additional names to match the number of arguments.
+   Unsetting the option will _____________, behavior that is more
```
"goal", as you put it, is better than "conclusion" here, eh?

Zimmi48 replied:
Indeed, but when I was recommending the use of "conclusion", this is not what I meant (as explained above): the conclusion is the part of the goal that is after the last `->`.

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+   |   conj : A -> B -> A /\\ B
+   | where "A /\\ B" := (and A B) : type_scope.
+
+   **quoted block above isn't formatted as I wanted.  What's the right way to get output to exactly match the input?**
```
I had tried `::` and indentation but Sphinx seemed to get confused by the indented lines within.  `coqdoc` works better, although it gives an unneed blank line at the top.

jfehrle commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+  of :n:`let-ins`, the patterns are applied to the arguments and :n:`let-in` variables.
+
+* :n:`( {+& @simple_intropattern} )` - matches a term that consists of one or more
+  right-hand nested binary inductive types in a form such as :g:`(a1 OP (a2 OP ...))`.
```
OK.  Though generally "right-associative" is used to refer to a property of the operator rather than how the expression is nested.

Zimmi48 replied:
Oh, right. It's just that the usual case is with non-parenthesized expressions, in which case it will be nested that way only if the operators are right-associative. But your previous sentence made sense, indeed!

jfehrle commented:
> Ping: this intro-pattern should be removed from the grammar.

Do you mean remove it from the doc or the .mlg file?

jfehrle commented:
I made changes addressing your points, which I put in a separate commit (to be squashed).  I left some comments perhaps worth perusing and one or two questions in the unresolved conversations.  So just a few more details before we can merge.

Zimmi48 commented:
>Do you mean remove it from the doc or the .mlg file?

From the doc only. It's just a special (useless) case of the generic `(,)` form.

>I made changes addressing your points, which I put in a separate commit (to be squashed).

Thanks! I'll have a look.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -74,29 +74,31 @@ specified, the default selector is used.
-  .. productionlist:: `bindings_list`
+  .. productionlist:: bindings_list
+     ref           : ident
+                   : INT
```
Change this in:

```rst
  .. productionlist:: bindings_list
     ref           : `ident`
                   : `num`
```

`num` and `ident` are both already defined grammar tokens.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -168,29 +169,22 @@ The most common splitting patterns are:
-  :ref:`Example <intropattern_conj_ex>`
+  :g:`(H1, H2)` or :g:`[H1 H2]`.
+  :ref:`Example <intropattern_conj_ex>`.  This also works on :n:`A <-> B`, which
+  is just a notation representing :n:`(A -> B) /\ (B -> A)`.
```
Correct. But note that there were discussions of creating a proper inductive type for `<->` in which case this splitting pattern would still work. That's the magic :smile:

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -202,28 +196,25 @@ For a goal :g:`A \/ B`, use :tacn:`left` to replace the current goal with :g:`A`
 
 * :n:`( {+& @simple_intropattern} )` - matches a term that consists of one or more
-  right-hand nested binary inductive types in a form such as :g:`(a1 OP (a2 OP ...))`.
+  nested binary inductive types in a right-associative form such as :g:`(a1 OP (a2 OP ...))`.
```
Maybe keep your old formulation but instead remove the parentheses from your example and add "(supposing that OP is right-associative)"...

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -202,28 +196,25 @@ For a goal :g:`A \/ B`, use :tacn:`left` to replace the current goal with :g:`A`
   :ref:`Example <intropattern_ampersand_ex>`
 
-* :n:`[ {+| @intropatterns} ]` - splits an inductive type that has
+* :n:`[ {*| @intropatterns} ]` - splits an inductive type that has
```
Here we go. Now that you have changed `+` in `*`, the bullet-point above is just a specialization of this case.

jfehrle replied:
Not quite.  Both the 2 preceding points should have `+`.

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -202,28 +196,25 @@ For a goal :g:`A \/ B`, use :tacn:`left` to replace the current goal with :g:`A`
   :ref:`single constructor with multiple parameters <intropattern_cons_note>`
   such as :n:`A /\ B` into multiple hypotheses.  Use :n:`[H1 [H2 H3]]` to match :g:`A /\ B /\ C`.
-* :n:`[]` - splits an inductive type:  If the inductive type has multiple constructors,
+* :n:`[]` - splits an inductive type (a special case of the last 2 bullets):  If the inductive
```
On the contrary, I would not say that it is a "special case". It has a peculiar semantics...

Zimmi48 commented on doc/sphinx/proof-engine/tactics.rst:
```diff
@@ -456,6 +420,37 @@ Examples:
+         Theorem length_zero_iff_nil (l : list A):
+            length l = 0 <-> l=nil.
+
+      The example is based on https://github.com/tchajed/coq-tricks/blob/master/src/IntroPatterns.v
```
Very last comment: I suggest using a permalink, like https://github.com/tchajed/coq-tricks/blob/8e6efe4971ed828ac8bdb5512c1f615d7d62691e/src/IntroPatterns.v. Since it's a bit long, maybe this can become ``` `Tej Chajed's coq-tricks
<https://github.com/tchajed/coq-tricks/blob/8e6efe4971ed828ac8bdb5512c1f615d7d62691e/src/IntroPatterns.v>`_ ```

jfehrle commented:
@Zimmi48 I made changes addressing your comments except for the one about "Note that there is no equivalent to intro patterns for the conclusion of goals." where I didn't understand you.

I was travelling last week.  I didn't have the mental energy to respond after skiing all day, otherwise I would have responded sooner.  I think we're very close to done.

jfehrle commented:
Squashed, ready to merge.

jfehrle commented:
Yay!  Thanks for your help, Théo.

On Thu, Jan 24, 2019, 7:36 AM Théo Zimmermann <notifications@github.com
wrote:

> Merged #9269 <https://github.com/coq/coq/pull/9269> into master.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/coq/coq/pull/9269#event-2094815443>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/ABMf3UDEDLngq2NtHM6Eez-Pt6jdtwfcks5vGdMYgaJpZM4ZfLX5>
> .
>

Zimmi48 commented:
You're quite welcome. Thanks for the time you devoted to this!
